### ISSUE TITLE
error[invalid-argument-type] - Expected `int | float`, found `object`

### ISSUE THREAD

**Issue by gothicVI** [astral-sh/ty#3091]
### Summary

This code passes `mypy`, `pyright`, and `pyrefly`:
```python3
from collections.abc import Iterable
from typing import overload

import numpy as np

@overload
def convert_units_dec(arg: Iterable[int | float], /, *, dec: int = 2) -> list[str]: ...

@overload
def convert_units_dec(arg: int | float, /, *, dec: int = 2) -> str: ...

def convert_units_dec(arg: Iterable[int | float] | int | float, /, *, dec: int = 2) -> list[str] | str:

    def convert(i: int | float, /, *, dec: int = dec) -> str:
        if np.isnan(i):
            return "N/A"
        if i >= 1.0e15:
            return f"{np.round(i / 1.0e15, decimals=dec):.{dec}f}PB"
        if i >= 1.0e12:
            return f"{np.round(i / 1.0e12, decimals=dec):.{dec}f}TB"
        if i >= 1.0e9:
            return f"{np.round(i / 1.0e9, decimals=dec):.{dec}f}GB"
        if i >= 1.0e6:
            return f"{np.round(i / 1.0e6, decimals=dec):.{dec}f}MB"
        if i >= 1.0e3:
            return f"{np.round(i / 1.0e3, decimals=dec):.{dec}f}KB"
        return f"{int(np.round(i, decimals=0))}B"

    if isinstance(arg, Iterable):
        return [convert(i) for i in arg]
    return convert(arg)
```
but raises
```
error[invalid-argument-type]: Argument to function `convert` is incorrect
   |
   |     if isinstance(arg, Iterable):
   |         return [convert(i) for i in arg]
   |                         ^ Expected `int | float`, found `object`
   |     return convert(arg)
   |
info: Function defined here
   |
   |     def convert(i: int | float, /, *, dec: int = dec) -> str:
   |         ^^^^^^^ -------------- Parameter declared here
   |
info: rule `invalid-argument-type` is enabled by default
```

### Version

ty 0.0.24

**Comment by sharkdp** [comment]
Please see https://docs.astral.sh/ty/reference/typing-faq/#what-is-toplistunknown-and-why-does-it-appear and https://docs.astral.sh/ty/features/type-system/#top-and-bottom-materializations. For your case, replace `list` with `Iterable`, and note that `Iterable` is covariant in its type parameter, so the top materialization of `Iterable[Any]` is `object`.

And please let us know if it remains unclear, this is a complicated topic.

**Comment by sharkdp** [comment]
To fix it, you could rewrite it as
```py
    if isinstance(arg, int | float):
        return convert(arg)
    return [convert(i) for i in arg]
```

**Comment by sharkdp** [comment]
If you think this is too pedantic (the fact that ty considers the possibility of there being a common subtype of `int` and `Iterable` that would pass your `isinstance(…, Iterable)` check), see https://github.com/astral-sh/ty/issues/1578

**Comment by gothicVI** [comment]
Thanks - especially for the explanation.
